property/index.js

```diff
@@ -115,54 +119,133 @@ module.exports = yeoman.generators.Base.extend({
+        try {
+          coerceDefaultValue(this.propDefinition, answers.defaultValue);
+          debug('property definition: %j', this.propDefinition);
+        } catch (err) {
+          debug('Failed to coerce property default value: ', err);
+          this.log('Warning: please enter the ' + this.name +
+            ' property again. The default value provided "' +
+            answers.defaultValue + 
+            '" is not valid for the selected type: ' + this.type);
+          return this.askForParameters();
+        }
+      }
       done();
     }.bind(this));
   },
 
   property: function() {
     var done = this.async();
-    var def = {
-      name: this.name,
-      type: this.type
-    };
-    if (this.required) {
-      def.required = true;
-    }
-    if (this.defaultValue) {
-      if (this.type === 'boolean'){
-        if (['true', '1', 't'].indexOf(this.defaultValue) !== -1 ){
-          def.default = true;
-        } else {
-          def.default = false;
-        }
-      } else if (this.defaultValue === 'uuid' || this.defaultValue === 'guid'){
-         def.defaultFn = this.defaultValue;
-      } else if ((this.type === 'date' || this.type === 'datetime') &&
-                  this.defaultValue.toLowerCase() === 'now'){
-         def.defaultFn = 'now';
-      } else {
-         def.default = this.defaultValue;
-      }
-    }
-
-    this.modelDefinition.properties.create(def, function(err) {
+    this.modelDefinition.properties.create(this.propDefinition, function(err) {
       helpers.reportValidationError(err, this.log);
       return done(err);
     }.bind(this));
   },
-
   saveProject: actions.saveProject
 });
+
+function coerceDefaultValue(propDef, value) {
+  var itemType;
+  if (Array.isArray(propDef.type)) {
+    itemType = propDef.type[0];
+    propDef.type = 'array';
+  }
+
+  switch (propDef.type) {
+    case 'string':
+      if (value === 'uuid' || value === 'guid'){
+        propDef.defaultFn = value;
+      } else {
+        propDef.default = value;
+      }
+      break;
+    case 'number':
+      propDef.default = castToNumber(value);
+      break;
+    case 'boolean':
+      if (['true', '1', 't'].indexOf(value) !== -1 ){
+        propDef.default = true;
+      } else {
+        propDef.default = false;
+      }
+      break;
+    case 'date':
+      if (value.toLowerCase() === 'now'){
+        propDef.defaultFn = 'now';
+      } else {
+        var isNumber = /^[0-9]+$/.test(value);
+        if (isNumber) {
+          propDef.default = new Date(castToNumber(value));
+        } else {
+          propDef.default = new Date(value);
+        }
+      }
+      break;
+    case 'array':
+      propDef.type = [itemType];
+      if (itemType === 'string') {
+        propDef.default = value.replace(/[\s,]+/g, ',').split(',');
+      } else if (itemType === 'number') {
+        propDef.default = value.replace(/[\s,]+/g, ',').split(',')
+          .map(function(item) {
+            return Number(castToNumber(item));
```

Comment:
Why is it necessary to wrap the result of `castToNumber` in `Number` constructor one more time? The name `castToNumber` suggests it should already return the correct type, and the implementation of `castToNumber` does that.

I think this should be enough here: 

```js
value.replace(...).split(',')
  .map(castToNumber);
```